libafl/src/mutators/mapping.rs

```diff
@@ -0,0 +1,335 @@
+//! Allowing mixing and matching between [`Mutator`] and [`crate::inputs::Input`] types.
+use alloc::{borrow::Cow, vec::Vec};
+
+use libafl_bolts::{tuples::MappingFunctor, Named};
+
+use crate::{
+    inputs::MutVecInput,
+    mutators::{MutationResult, Mutator},
+    Error,
+};
+
+/// Mapping [`Mutator`] that allows using [`Mutator`]s for [`Vec<u8>`] on (parts of) other input types that can be mapped to [`Vec<u8>`].
+///
+/// # Example
+#[cfg_attr(feature = "std", doc = " ```")]
+#[cfg_attr(not(feature = "std"), doc = " ```ignore")]
+///
+/// use std::vec::Vec;
+///
+/// use libafl::{
+///     mutators::{
+///         ByteIncMutator, MutationResult, MutVecMappingMutator, Mutator,
+///     },
+///     state::NopState,
+/// };
+///
+/// type CustomInput = Vec<u8>;
+///
+/// let inner = ByteIncMutator::new();
+/// let mut outer = MutVecMappingMutator::new(inner);
+///
+/// let mut input: CustomInput = vec![1];
+/// let mut state: NopState<CustomInput> = NopState::new();
+/// let res = outer.mutate(&mut state, &mut input).unwrap();
+/// assert_eq!(res, MutationResult::Mutated);
+/// assert_eq!(input, vec![2]);
+/// ```
+#[derive(Debug)]
+pub struct MutVecMappingMutator<M> {
+    inner: M,
+}
+
+impl<M> MutVecMappingMutator<M> {
+    /// Creates a new [`MutVecMappingMutator`]
+    pub fn new(inner: M) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S, M> Mutator<Vec<u8>, S> for MutVecMappingMutator<M>
```

Comment:
Instead of `Vec<u8>` this dude should operate on `HasMutatorBytes`, then it'll work on whatever input you pass in

Reply:
I don't think `HasMutatorBytes` is implemented for `Vec<u8>`, this is the exact reason `MutVecMappingMutator` is even necessary.

Reply:
It's not implemented for Vec<u8> but it's implemented for `MutVecInput` and `MutVecInput`, in turn, is implemented for `&mut Vec<u8>`, so I think you can easily do this without hard-coding a `Vec<u8>`

Reply:
This is exactly what `MutVecMappingMutator` does. I just figured it'd be one more step every user of `havoc_mutations_mapped` would need to do themselves, and it'd be required basically always, and it can be centrally done, so I opted to include this logic in the library.